relay-server/src/actors/store.rs

```diff
@@ -199,51 +199,101 @@ impl StoreService {
             }
         }
 
-        if let Some(event_item) = event_item {
-            relay_log::trace!("Sending event item of envelope to kafka");
-            let event_message = KafkaMessage::Event(EventKafkaMessage {
-                payload: event_item.payload(),
-                start_time: UnixTimestamp::from_instant(start_time).as_secs(),
-                event_id: event_id.ok_or(StoreError::NoEventId)?,
-                project_id: scoping.project_id,
-                remote_addr: envelope.meta().client_addr().map(|addr| addr.to_string()),
-                attachments,
-            });
+        if event_item.is_none() && attachments.is_empty() {
+            return Ok(());
+        }
 
-            self.produce(topic, scoping.organization_id, event_message)?;
-            metric!(
-                counter(RelayCounters::ProcessingMessageProduced) += 1,
-                event_type = &event_item.ty().to_string()
-            );
-        } else if !attachments.is_empty() {
-            relay_log::trace!("Sending individual attachments of envelope to kafka");
-            for attachment in attachments {
-                let attachment_message = KafkaMessage::Attachment(AttachmentKafkaMessage {
-                    event_id: event_id.ok_or(StoreError::NoEventId)?,
-                    project_id: scoping.project_id,
-                    attachment,
-                });
+        let remote_addr = envelope.meta().client_addr().map(|addr| addr.to_string());
 
-                self.produce(topic, scoping.organization_id, attachment_message)?;
-                metric!(
-                    counter(RelayCounters::ProcessingMessageProduced) += 1,
-                    event_type = "attachment"
-                );
+        let kafka_messages = Self::extract_kafka_messages(
+            event_item,
+            event_id.ok_or(StoreError::NoEventId)?,
+            scoping,
+            start_time,
+            remote_addr,
+            attachments,
+        );
+
+        for message in kafka_messages {
+            if let Ok(()) = self.produce(topic, scoping.organization_id, &message) {
+                // If message was succesfully sent, make the right log message.
+                match message {
+                    KafkaMessage::Attachment(_) => {
+                        metric!(
+                            counter(RelayCounters::ProcessingMessageProduced) += 1,
+                            event_type = "attachment"
+                        );
+                    }
+                    _ => {
+                        if let Some(event_item) = event_item {
+                            metric!(
+                                counter(RelayCounters::ProcessingMessageProduced) += 1,
+                                event_type = &event_item.ty().to_string()
+                            );
+                        }
+                    }
+                }
             }
         }
 
         Ok(())
     }
 
+    fn extract_kafka_messages(
```

Comment:
To clarify that this is only for envelopes with an event ID:

```suggestion
    fn extract_kafka_messages_for_event(
```